ads/google/a4a/utils.js

```diff
@@ -375,56 +375,75 @@ export function extractAmpAnalyticsConfig(a4a, responseHeaders) {
-      // TODO: Ensure url is valid and not freeform JS?
-      requests[`visibility${idx}`] = `${urls[idx - 1]}`;
-    }
-    // Security review needed here.
-    config['requests'] = requests;
-    config['triggers']['continuousVisible']['request'] =
-        Object.keys(requests);
-    // Add CSI pingbacks.
-    const correlator = getCorrelator(a4a.win);
-    const slotId = a4a.element.getAttribute('data-amp-slot-index');
     const qqid = (responseHeaders && responseHeaders.has(QQID_HEADER))
-        ? responseHeaders.get(QQID_HEADER) : 'null';
-    config['requests']['visibilityCsi'] =
-        'https://csi.gstatic.com/csi?fromAnalytics=1' +
-        `&c=${correlator}&slotId=${slotId}&qqid.0=${qqid}`;
-    config['triggers']['continuousVisibleIniLoad']['request'] =
-        'visibilityCsi';
-    config['triggers']['continuousVisibleRenderStart']['request'] =
-        'visibilityCsi';
-    return config;
+      ? responseHeaders.get(QQID_HEADER) : 'null';
+    return buildAmpAnalyticsConfig(a4a.win, a4a.element, urls, true, qqid);
+
   } catch (err) {
     dev().error('AMP-A4A', 'Invalid analytics', err,
         responseHeaders.get(AMP_ANALYTICS_HEADER));
   }
   return null;
 }
+
+/**
+ * Builds the Amp Analytics configuration object.
+ *
+ * @param {!Window} win
+ * @param {!element} element
+ * @param {!Array.<string>} urls Array of urls to send pings to.
+ * @param {!boolean} isA4a
+ * @param {string=} opt_qqid
+ * @return {?JSONType} config
+ */
+export function buildAmpAnalyticsConfig(win, element, urls, isA4a, opt_qqid) {
+  const config = /** @type {JSONType}*/ ({
+    'transport': {'beacon': false, 'xhrpost': false},
+    'triggers': {
+      'continuousVisible': {
+        'on': 'visible',
+        'visibilitySpec': {
+          'selector': 'amp-ad',
+          'selectionMethod': 'closest',
+          'visiblePercentageMin': 50,
+          'continuousTimeMin': 1000,
+        },
+      },
+      'continuousVisibleIniLoad': {
+        'on': 'ini-load',
+        'selector': 'amp-ad',
+        'selectionMethod': 'closest',
+      },
+      'continuousVisibleRenderStart': {
+        'on': 'render-start',
+        'selector': 'amp-ad',
+        'selectionMethod': 'closest',
+      },
+    },
+  });
+  const requests = {};
+  for (let idx = 1; idx <= urls.length; idx++) {
+    // TODO: Ensure url is valid and not freeform JS?
+    requests[`visibility${idx}`] = `${urls[idx - 1]}`;
+  }
+  // Security review needed here.
+  config['requests'] = requests;
+  config['triggers']['continuousVisible']['request'] =
+      Object.keys(requests);
+  // Add CSI pingbacks.
+  // TODO(bradfrizzell@) talk to tdrl@ about this
+  const correlator = getCorrelator(win);
+  const slotId = element.getAttribute('data-amp-slot-index');
+
+  config['requests']['visibilityCsi'] =
+      'https://csi.gstatic.com/csi?fromAnalytics=1' +
+      `&c=${correlator}&slotId=${slotId}` +
+      `&a4a=` + (isA4a ? '2' : '1') +
```

Comment:
Let's invert this an only include when true (e.g. df=1 only when delayed fetch)
Can we also add the type so we can differentiate adsense and doubleclick